### ISSUE TITLE
Static linking with `libopus.a` and `libcelt0.a`

### ISSUE THREAD

**Issue by msink** [msys2/MINGW-packages#11515]
If both `libopus.a` and `libcelt0.a` have to be linked in application - linker error:

```
...\lib\libopus.a(celt_encoder.o):(.text+0x830): multiple definition of `celt_encoder_get_size'; ...\lib\libcelt0.a(celt.o):(.text+0x4bb0): first defined here
...\lib\libopus.a(celt_encoder.o):(.text+0x6940): multiple definition of `celt_encoder_init'; ...\lib\libcelt0.a(celt.o):(.text+0x4d50): first defined here
...\lib\libopus.a(mdct.o):(.text+0x0): multiple definition of `clt_mdct_init'; ...\lib\libcelt0.a(mdct.o):(.text+0x0): first defined here
...\lib\libopus.a(mdct.o):(.text+0x180): multiple definition of `clt_mdct_clear'; ...\lib\libcelt0.a(mdct.o):(.text+0x110): first defined here
...\lib\libopus.a(pitch.o):(.text+0x2c0): multiple definition of `pitch_downsample'; ...\lib\libcelt0.a(pitch.o):(.text+0x170): first defined here
...\lib\libopus.a(pitch.o):(.text+0x620): multiple definition of `pitch_search'; ...\lib\libcelt0.a(pitch.o):(.text+0x3c0): first defined here
...\lib\libopus.a(quant_bands.o):(.text+0x5b0): multiple definition of `quant_coarse_energy'; ...\lib\libcelt0.a(quant_bands.o):(.text+0x500): first defined here
...\lib\libopus.a(quant_bands.o):(.text+0xc10): multiple definition of `quant_fine_energy'; ...\lib\libcelt0.a(quant_bands.o):(.text+0xa20): first defined here
...\lib\libopus.a(quant_bands.o):(.text+0xe50): multiple definition of `quant_energy_finalise'; ...\lib\libcelt0.a(quant_bands.o):(.text+0xbf0): first defined here
...\lib\libopus.a(quant_bands.o):(.text+0x1090): multiple definition of `unquant_coarse_energy'; ...\lib\libcelt0.a(quant_bands.o):(.text+0xdd0): first defined here
...\lib\libopus.a(quant_bands.o):(.text+0x1370): multiple definition of `unquant_fine_energy'; ...\lib\libcelt0.a(quant_bands.o):(.text+0x1070): first defined here
...\lib\libopus.a(quant_bands.o):(.text+0x1480): multiple definition of `unquant_energy_finalise'; ...\lib\libcelt0.a(quant_bands.o):(.text+0x1180): first defined here
...\lib\libopus.a(quant_bands.o):(.text+0x1650): multiple definition of `amp2Log2'; ...\lib\libcelt0.a(quant_bands.o):(.text+0x14d0): first defined here
...\lib\libopus.a(rate.o):(.text+0x0): multiple definition of `compute_pulse_cache'; ...\lib\libcelt0.a(rate.o):(.text+0x0): first defined here
...\lib\libopus.a(bands.o):(.text+0x1690): multiple definition of `compute_band_energies'; ...\lib\libcelt0.a(bands.o):(.text+0x2b0): first defined here
...\lib\libopus.a(bands.o):(.text+0x17b0): multiple definition of `normalise_bands'; ...\lib\libcelt0.a(bands.o):(.text+0x420): first defined here
...\lib\libopus.a(bands.o):(.text+0x18a0): multiple definition of `denormalise_bands'; ...\lib\libcelt0.a(bands.o):(.text+0x510): first defined here
...\lib\libopus.a(bands.o):(.text+0x1ac0): multiple definition of `anti_collapse'; ...\lib\libcelt0.a(bands.o):(.text+0x670): first defined here
...\lib\libopus.a(bands.o):(.text+0x1f20): multiple definition of `spreading_decision'; ...\lib\libcelt0.a(bands.o):(.text+0xab0): first defined here
...\lib\libopus.a(bands.o):(.text+0x2260): multiple definition of `haar1'; ...\lib\libcelt0.a(bands.o):(.text+0xd80): first defined here
...\lib\libopus.a(bands.o):(.text+0x2ff0): multiple definition of `quant_all_bands'; ...\lib\libcelt0.a(bands.o):(.text+0x2e00): first defined here
...\lib\libopus.a(celt_decoder.o):(.text+0x1630): multiple definition of `celt_decoder_get_size'; ...\lib\libcelt0.a(celt.o):(.text+0x5280): first defined here
...\lib\libopus.a(celt_decoder.o):(.text+0x3250): multiple definition of `celt_decoder_init'; ...\lib\libcelt0.a(celt.o):(.text+0x53c0): first defined here
...\lib\libopus.a(celt_lpc.o):(.text+0x0): multiple definition of `_celt_lpc'; ...\lib\libcelt0.a(plc.o):(.text+0x0): first defined here
...\lib\libopus.a(celt_lpc.o):(.text+0x6b0): multiple definition of `_celt_autocorr'; ...\lib\libcelt0.a(plc.o):(.text+0x3a0): first defined here
```
To repoduce: see https://github.com/msys2/MINGW-packages/issues/8824#issuecomment-1087561620
replace `-l:libcelt0.dll.a` to `-lcelt0`

**Comment by Biswa96** [comment]
Are libopus.a and libcelt0.a meant to be linked together?

**Comment by msink** [comment]
Dont't know, but if in https://github.com/msys2/MINGW-packages/issues/8824#issuecomment-1087561620 remove `-l:libcelt0.dll.a`:
```
...\lib\libavcodec.a(libcelt_dec.o):libcelt_dec.c:(.text+0x4f): undefined reference to `celt_decode'
...\lib\libavcodec.a(libcelt_dec.o):libcelt_dec.c:(.text.unlikely+0xe): undefined reference to `celt_decoder_destroy'
...\lib\libavcodec.a(libcelt_dec.o):libcelt_dec.c:(.text.unlikely+0x16): undefined reference to `celt_mode_destroy'
...\lib\libavcodec.a(libcelt_dec.o):libcelt_dec.c:(.text.unlikely+0x76): undefined reference to `celt_mode_create'
...\lib\libavcodec.a(libcelt_dec.o):libcelt_dec.c:(.text.unlikely+0x8f): undefined reference to `celt_decoder_create_custom'
...\lib\libavcodec.a(libcelt_dec.o):libcelt_dec.c:(.text.unlikely+0xa0): undefined reference to `celt_mode_destroy'
...\lib\libavcodec.a(libcelt_dec.o):libcelt_dec.c:(.text.unlikely+0x129): undefined reference to `celt_header_init'
```

**Comment by Biswa96** [comment]
Even libopus-dev Debian package has celt bundled in static library.

```
$ ar t /usr/lib/x86_64-linux-gnu/libopus.a  | grep celt
celt.o
celt_encoder.o
celt_decoder.o
celt_lpc.o
x86_celt_map.o
celt_lpc_sse4_1.o
```

It is not impossible to replace the bundled celt library in opus package. But I think it would be better not to fiddle with already a stable package and default upstream choice.

**Comment by Biswa96** [comment]
I have asked this in opus IRC and here is the discussion.

```
Apr 18 18:06:53 <Biswa96>	Is there any option to disable bundled celt library while building opus?
Apr 18 18:16:37 <kepstin>	not sure what you mean. opus is significantly based on celt; that's not really a bundled celt library but rather a core piece of the opus codec.
Apr 18 18:24:35 <Biswa96>	Someone wants to do  static linking and this happens https://github.com/msys2/MINGW-packages/issues/11515
Apr 18 18:24:46 <kepstin>	I don't think the older standalone celt libraries are compatible with opus anyways - or at very least are missing several years worth of significant bugfixes/improvement - and the level of integration means that you'd probably run into linking issues regardless.
Apr 18 18:25:22 <kepstin>	hmm, that's an interesting problem
Apr 18 18:25:40 <kepstin>	the "easy" solution is to disable the separate celt codec if it's not needed
Apr 18 18:26:23 <kepstin>	i guess this is only a problem with static linking, since the celt symbols in libopus would be private when doing dynamic linking?
Apr 18 18:28:35 <Biswa96>	Yes, the issue is only for static linking.
```

tl;dr static linking ❌

**Comment by msink** [comment]
So - static FFmpeg is impossible, and never will?

If `libopus` have embedded `libcelt` - ok, can it export functions from https://github.com/msys2/MINGW-packages/issues/11515#issuecomment-1101354803 so FFmpeg's `libavcodec.a` can use them instead?

**Comment by kepstin** [comment]
My suggestion would be to disable the celt codec when building static ffmpeg, and leave opus enabled. Standalone celt is extremely niche, while opus is in common use.

**Comment by Biswa96** [comment]
Yeah, that's another option. ArchLinux does not use celt with anything.
